### PR TITLE
New method to get variables values

### LINKED ISSUE #3418: ValidationContext.GetVariableValuesAsync has side effects
Today I faced one funny bug (side effect) in one of our services. The same validation error was listed 5 times:
изображение

This is "be design" of `ValidationContext.GetVariableValuesAsync`. This method parses inputs and coerces them into appropriate types. **Also** it reports validation errors via `ReportError` call (4 calls in ValidationContext.cs). OK? Yes... but this method is public and one of our custom validation rules calls it as well to get variables list. This rule does so inside `MathingNodeVisitor<GraphQLField>` so each time traversing algorithm visits some field `ValidationContext.GetVariableValuesAsync` method is called. In case of invalid variable **each** field in query yields new element in `ValidationContext.Errors` about the same problem. 5 fields -> 5 errors, 100 fields -> 100 errors. I think this is a design flaw. We should not add errors **inside** this method. Instead we should return them as a result along with variables. The only place where we use `GetVariableValuesAsync` is `DocumentValidator` just before returning `ValidationResult` to the called. The code easily may be rewritten to append returned collection of errors to existing errors in `context` variable:
изображение

But... it's a breaking change for `GetVariableValuesAsync`. Thoughts?

The current workaround for me now - make a clone of `ValidationContext` instance in our validation rule each time I want to call `GetVariableValuesAsync`.

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #3418 

I just copied `GetVariableValuesAsync` body with little changes to return errors.

### CODE REVIEW

File: src/GraphQL/Validation/ValidationContext.cs [hunk 1]
```diff
@@ -128,7 +128,10 @@ public List<GraphQLFragmentSpread> GetFragmentSpreads(GraphQLSelectionSet node)
 
         /// <summary>
         /// Returns all of the variable values defined for the operation from the attached <see cref="Variables"/> object.
+        /// Only correctly validated variables are returned. If the variable is specified incorrectly, then an instance of
+        /// <see cref="ValidationError"/> is added to the <see cref="Errors"/>.
         /// </summary>
+        [Obsolete("This method can repeatedly add the same error into ValidationContext.Errors collection when called multiple times. Use GetVariablesValuesAsync method instead.")]
         public async ValueTask<Variables> GetVariableValuesAsync(IVariableVisitor? visitor = null)
```
Comment (Shane32) [inline src/GraphQL/Validation/ValidationContext.cs:135]: Can this call the `GetVariablesValuesAsync` method?

Comment (sungam3r) [reply src/GraphQL/Validation/ValidationContext.cs:135]: yes)

**Comment by Shane32** [comment]
I don't mind this PR as a temporary patch for 7.x, but I don't like this is as a long-term solution to the underlying issue.

**Comment by sungam3r** [comment]
What do you suggest?

Actually I need a patch for v5.

File: src/GraphQL/Validation/ValidationContext.cs [hunk 2]
```diff
@@ -128,16 +128,38 @@ public List<GraphQLFragmentSpread> GetFragmentSpreads(GraphQLSelectionSet node)
...
+
+            return variables;
+        }
+
+        /// <summary>
+        /// Returns all of the variable values defined for the operation from the attached <see cref="Variables"/> object.
+        /// Only correctly validated variables are returned. If the variable is specified incorrectly, then an instance of
+        /// <see cref="ValidationError"/> is returned within the list of errors.
+        /// </summary>
+        public async ValueTask<(Variables variables, List<ValidationError>? errors)> GetVariablesValuesAsync(IVariableVisitor? visitor = null)
```
Comment (Shane32) [inline src/GraphQL/Validation/ValidationContext.cs:153]:
```suggestion
        public async ValueTask<(Variables Variables, List<ValidationError>? Errors)> GetVariablesValuesAsync(IVariableVisitor? visitor = null)
```

File: src/GraphQL.Benchmarks/Benchmarks/DetailedBenchmark.cs [hunk 3]
```diff
@@ -223,7 +223,7 @@ public Validation.Variables ParseVariables()
                 Schema = Schema,
                 Variables = Inputs ?? Inputs.Empty,
                 Operation = Operation,
-            }.GetVariableValuesAsync().Result;
+            }.GetVariablesValuesAsync().Result.variables;
```
Comment (Shane32) [inline src/GraphQL.Benchmarks/Benchmarks/DetailedBenchmark.cs:226]:
```suggestion
            }.GetVariablesValuesAsync().Result.Variables;
```

File: src/GraphQL.ApiTests/net6/GraphQL.approved.txt [hunk 4]
```diff
@@ -3036,9 +3036,15 @@ namespace GraphQL.Validation
...
         public System.Collections.Generic.List<GraphQLParser.AST.GraphQLFragmentDefinition>? GetRecursivelyReferencedFragments(GraphQLParser.AST.GraphQLOperationDefinition operation) { }
         public System.Collections.Generic.List<GraphQLParser.AST.GraphQLFragmentDefinition>? GetRecursivelyReferencedFragments(System.Collections.Generic.List<GraphQLParser.AST.GraphQLOperationDefinition> operations) { }
+        [System.Obsolete("This method can repeatedly add the same error into ValidationContext.Errors colle" +
+            "ction when called multiple times. Use GetVariablesValuesAsync method instead.")]
         public System.Threading.Tasks.ValueTask<GraphQL.Validation.Variables> GetVariableValuesAsync(GraphQL.Validation.IVariableVisitor? visitor = null) { }
         public System.Collections.Generic.List<GraphQL.Validation.VariableUsage>? GetVariables<TNode>(TNode node)
             where TNode : GraphQLParser.AST.ASTNode, GraphQLParser.AST.IHasSelectionSetNode { }
+        [return: System.Runtime.CompilerServices.TupleElementNames(new string?[]?[] {
+                "variables",
+                "errors"})]
```
Comment (Shane32) [inline src/GraphQL.ApiTests/net6/GraphQL.approved.txt:3046]:
```suggestion
                "Variables",
                "Errors"})]
```

File: src/GraphQL.ApiTests/net5/GraphQL.approved.txt [hunk 5]
```diff
@@ -3022,9 +3022,15 @@ namespace GraphQL.Validation
...
         public System.Collections.Generic.List<GraphQLParser.AST.GraphQLFragmentDefinition>? GetRecursivelyReferencedFragments(GraphQLParser.AST.GraphQLOperationDefinition operation) { }
         public System.Collections.Generic.List<GraphQLParser.AST.GraphQLFragmentDefinition>? GetRecursivelyReferencedFragments(System.Collections.Generic.List<GraphQLParser.AST.GraphQLOperationDefinition> operations) { }
+        [System.Obsolete("This method can repeatedly add the same error into ValidationContext.Errors colle" +
+            "ction when called multiple times. Use GetVariablesValuesAsync method instead.")]
         public System.Threading.Tasks.ValueTask<GraphQL.Validation.Variables> GetVariableValuesAsync(GraphQL.Validation.IVariableVisitor? visitor = null) { }
         public System.Collections.Generic.List<GraphQL.Validation.VariableUsage>? GetVariables<TNode>(TNode node)
             where TNode : GraphQLParser.AST.ASTNode, GraphQLParser.AST.IHasSelectionSetNode { }
+        [return: System.Runtime.CompilerServices.TupleElementNames(new string?[]?[] {
+                "variables",
+                "errors"})]
```
Comment (Shane32) [inline src/GraphQL.ApiTests/net5/GraphQL.approved.txt:3032]:
```suggestion
                "Variables",
                "Errors"})]
```

**Review by Shane32 — CHANGES_REQUESTED** [review]

**Comment by Shane32** [comment]
> Actually I need a patch for v5.

We can backport this patch.

> What do you suggest?

Not sure offhand.  I would like to discuss it.  Here's one idea:

- New interface for validation rules that need to check variables.
- These rules run once standard validation rules have been run, and always run whether or not document caching is enabled.
- Field nodes are replaced with a derived class that also contains field arguments (using existing methods that read arguments and deserialize them, whether from variables or anything else).
- Then execution engine can use these field arguments instead of parsing them a second time.
- This has an added benefit of deserializing all argument literal values before any execution begins, which does not happen now, and could potentially throw an error, such as when passing an invalid date string to a DateTime field argument as a literal

**Comment by Shane32** [comment]
Building on that basis, we could perhaps add validation methods to fields and scalars.  So:

1. A scalar can inspect directives defined on the field to determine whether the field passes validation; and
2. During code-first type initialization, a simple call can set a validation rule on a field

Ideally could be 'compiled' during schema initialization so that only the minimum of calls are made.

**Comment by sungam3r** [comment]
It looks like a rather big redesign.

**Comment by Shane32** [comment]
> It looks like a rather big redesign.

It's not so bad.  See #3424 for the first part which moves argument parsing to the document validator.  Next PR could add ways to hook into the validation.

**Review by Shane32 — APPROVED** [review]

**Comment by sungam3r** [comment]
What is the process of patching v5.x ? I see 5.3.3 tag , origin/v533 branch, origin/v5 branch with merged (but not released) #3381 in it.

**Comment by sungam3r** [comment]
I suppose I need to modify all workflows to support `v5` branch.

**Comment by Shane32** [comment]
No. Just make a PR against the v5 branch. The one line to run the tests is already changed in that branch. No changes necessary to publish.

**Comment by sungam3r** [comment]
Even `build-artifacts-code.yml` to generate
изображение
?

**Comment by Shane32** [comment]
I think that's mainly used to publish beta builds.

**Pull request merged**